packages/workers-utils/src/compatibility-date.ts

```diff
@@ -0,0 +1,97 @@
+import assert from "node:assert";
+import module from "node:module";
+
+type YYYY = `${number}${number}${number}${number}`;
+type MM = `${number}${number}`;
+type DD = `${number}${number}`;
+
+/**
+ * Represents a valid compatibility date, a string such as `2025-09-27`
+ */
+export type CompatDate = `${YYYY}-${MM}-${DD}`;
+
+type GetCompatDateOptions = {
+	projectPath?: string;
+};
+
+type GetCompatDateResult = {
+	date: CompatDate;
+	source: "workerd" | "fallback" | "today";
+};
+
+/**
+ * Gets compatibility date either of the locally installed workerd package.
+ *
+ * If the package is not found the fallback date of 2025-09-27 is returned instead.
+ *
+ * Additionally, if the workerd date is set to the future then the current date is returned instead.
+ *
+ * @param options.projectPath the path to the project
+ * @returns an object including the compatibility date and its source
+ */
+export function getLocalWorkerdCompatibilityDate({
+	projectPath = process.cwd(),
+}: GetCompatDateOptions = {}): GetCompatDateResult {
+	try {
+		const projectRequire = module.createRequire(projectPath);
+		const miniflareEntry = projectRequire.resolve("miniflare");
+		const miniflareRequire = module.createRequire(miniflareEntry);
+		const miniflareWorkerd = miniflareRequire("workerd") as {
+			compatibilityDate: string;
+		};
+		const workerdDate = miniflareWorkerd.compatibilityDate;
+		return toSafeCompatDateObject({
+			date: formatCompatibilityDate(new Date(workerdDate)),
+			source: "workerd",
+		});
+	} catch {}
+
+	const fallbackDate = new Date("2025-09-27");
+	return toSafeCompatDateObject({
+		date: formatCompatibilityDate(fallbackDate),
```

Comment:
Do we need to create and then parse the date at runtime (could be in the tests if needed)
```suggestion
	return toSafeCompatDateObject({
		date: "2025-09-27",
```